### PR TITLE
DP-2459 api ref generator

### PR DESCRIPTION
**Pull request by Mara3l**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: scripts/docs/json_builder.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+            # Filter out non-gooddata libraries
+            if MODULE_NAME in obj_module.__name__:
+                data[name] = class_data(obj)
+        elif isinstance(obj, ModuleType):
+            if MODULE_NAME in obj_module.__name__:
+                data[name] = module_data(obj)
+    return data
+
+
+def parse_package(obj: ModuleType, data=None):
```
Comment (hkad98) [inline scripts/docs/json_builder.py:242]: data: Optional[dict] = None

File: scripts/docs/json_builder.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+            if item.ispkg:
+                data[item.name].update(parse_package(vars(obj)[item.name]))
+            else:
+                module = vars(obj)[item.name]
+                data[item.name].update(module_data(module))
+
+    return data
+
+
+def import_submodules(pkg_name):
```
Comment (hkad98) [inline scripts/docs/json_builder.py:281]: Missing typing.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: scripts/docs/json_builder.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+        iterator = pkgutil.iter_modules(obj.__path__)
+        for item in iterator:
+            if item.name not in data:
+                data[item.name] = {}
+
+            if item.ispkg:
+                data[item.name].update(parse_package(vars(obj)[item.name]))
+            else:
+                module = vars(obj)[item.name]
+                data[item.name].update(module_data(module))
```
Comment (hkad98) [inline scripts/docs/json_builder.py:277]: Please make the following change so we have better readability.

```python
if not isinstance(obj, ModuleType):
    return data
iterator = pkgutil.iter_modules(obj.__path__)
        for item in iterator:
            if item.name not in data:
                data[item.name] = {}

            if item.ispkg:
                data[item.name].update(parse_package(vars(obj)[item.name]))
            else:
                module = vars(obj)[item.name]
                data[item.name].update(module_data(module))

```

File: scripts/docs/python_ref_builder.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+import json
+from pathlib import Path
+from typing import List
+
+MODULE_TEMPLATE_STRING = Path("module_template.md").read_text()
+CLASS_TEMPLATE_STRING = Path("class_template.md").read_text()
+FUNCTION_TEMPLATE_STRING = Path("function_template.md").read_text()
+
+
+def process_json_file(file_path: str) -> dict:
```
Comment (hkad98) [inline scripts/docs/python_ref_builder.py:12]: read_json_file

File: scripts/docs/json_builder.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+        if data.get("kind", None) in ["class", "function"]:
+            # Last item in path corresponds to the class name
+            links[path.split(".")[-1]] = {"path": path, "kind": data["kind"]}
+
+        for key, val in data.items():
+            if isinstance(val, dict):
+                if key == "functions":
+                    _recursive_find_classes(val, path)
+                elif path != "":
+                    _recursive_find_classes(val, f"{path}.{key}")
```
Comment (hkad98) [inline scripts/docs/json_builder.py:324]: I do not fully understand this elif part? If `path != ""` then this branch is hit always, right? `

Can we do something like `path_key = f"{path}.{key}"`?

File: scripts/docs/json_builder.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+                else:
+                    _recursive_find_classes(val, key)
+
+    _recursive_find_classes(module_data, "")
```
Comment (hkad98) [inline scripts/docs/json_builder.py:328]: What does this?

File: scripts/docs/python_ref_builder.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+            # Create files based on the kind of the data: module/class/function
+            if kind == "module":
+                (dir_root / name).mkdir(exist_ok=True)
+                (dir_root / name / "_index.md").open("w+").write(
+                    MODULE_TEMPLATE_STRING.replace("PATH", obj_module_import_path)
+                    .replace("NAME", name)
+                    .replace("LINK", name)
+                )
+                links[name] = {"path": f"{url_root}/{name}".lower(), "kind": "function"}  # Lowercase for Hugo
```
Comment (hkad98) [inline scripts/docs/python_ref_builder.py:70]: It seems that some code is repeating. Create utility function that does the majority of the repeating code.

File: scripts/docs/python_ref_builder.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+                        .replace("PARENT", module_import_path.split(".")[-1])
+                    )
+                    links[func_name] = {
+                        "path": f"{url_root}/{func_name}".lower(),  # Lowercase for Hugo
+                        "kind": "function",
+                    }
+                continue  # No need to recurse deeper
+
+            else:
+                continue  # Not a class nor a module
```
Comment (hkad98) [inline scripts/docs/python_ref_builder.py:98]: What is here? A file?

File: scripts/docs/json_builder.py [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+
+
+if __name__ == "__main__":
+    import gooddata_sdk
+
+    MODULE_NAME = "gooddata_sdk"  # This global variable is needed in further parsing
+    import_submodules(MODULE_NAME)
+    res = parse_package(gooddata_sdk)
+    output_json: dict = cattrs.unstructure(res)
+    open("data.json", "w+").write(json.dumps(output_json))
```
Comment (Mara3l) [inline scripts/docs/json_builder.py:342]: Please. change this to w, as this would only append to it.

File: scripts/generate.sh [hunk 10]
```diff
@@ -69,6 +69,15 @@ for branch in "$remote_name/master" $(git branch -rl "$remote_name/rel/*") ; do
         git archive "$branch" "content/en/$src_section" | tar xf - -C "$content_dir/$target_section" \
             --strip-components=$strip_count "content/en/$src_section"
     fi
+    API_GEN_FILE="$branch:scripts/docs/json_builder.py"
+    if git cat-file -e $API_GEN_FILE; then
+        echo "$API_GEN_FILE existss."
```
Comment (Mara3l) [inline scripts/generate.sh:74]: Please fix typo in existss.

File: scripts/docs/python_ref_builder.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    )
+    parser.add_argument("--url_root", default="", required=False, help="url root path for the apiref")
+    args = parser.parse_args()
+    print(f"Json start path is f{args.json_start_path}")
+
+    file_path = args.file
+    data = process_json_file(file_path)
+    data = change_json_root(data, args.json_start_path)
+    links = create_file_structure(data, Path(args.output), url_root=args.url_root)
+    json.dump(links, open("links.json", "w+"), indent=4)
```
Comment (Mara3l) [inline scripts/docs/python_ref_builder.py:150]: This should also be w, right?

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Review by hkad98 — APPROVED** [review]

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-python-sdk-gate-PR316

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Pull request merged**